src/storage/sftp/write.c

```diff
@@ -0,0 +1,461 @@
+    StorageWriteSftp *
+#define FUNCTION_LOG_STORAGE_WRITE_SFTP_FORMAT(value, buffer, bufferSize)                                                          \
+    objNameToLog(value, "StorageWriteSftp", buffer, bufferSize)
+
+/***********************************************************************************************************************************
+Open the file
+***********************************************************************************************************************************/
+static void
+storageWriteSftpOpen(THIS_VOID)
+{
+    THIS(StorageWriteSftp);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(STORAGE_WRITE_SFTP, this);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->sftpSession != NULL);
+
+    const unsigned long int flags = LIBSSH2_FXF_CREAT | LIBSSH2_FXF_WRITE | (this->interface.truncate ? LIBSSH2_FXF_TRUNC : 0);
+
+    // Open the file
+    Wait *wait = waitNew(this->timeoutConnect);
+
+    do
+    {
+        this->sftpHandle = libssh2_sftp_open_ex(
+            this->sftpSession, strZ(this->nameTmp), (unsigned int)strSize(this->nameTmp), flags, (int)this->interface.modeFile,
+            LIBSSH2_SFTP_OPENFILE);
+    }
+    while (this->sftpHandle == NULL && libssh2_session_last_errno(this->session) == LIBSSH2_ERROR_EAGAIN && waitMore(wait));
+
+    // Attempt to create the path if it is missing
+    if (this->sftpHandle == NULL && this->interface.createPath)
+    {
+        // Create the path
+        storageInterfacePathCreateP(this->storage, this->path, false, false, this->interface.modePath);
+
+        // Open file again
+        wait = waitNew(this->timeoutConnect);
+
+        do
+        {
+            this->sftpHandle = libssh2_sftp_open_ex(
+                this->sftpSession, strZ(this->nameTmp), (unsigned int)strSize(this->nameTmp), flags, (int)this->interface.modeFile,
+                LIBSSH2_SFTP_OPENFILE);
+        }
+        while (this->sftpHandle == NULL && (libssh2_session_last_errno(this->session) == LIBSSH2_ERROR_EAGAIN && waitMore(wait)));
+    }
+
+    // Handle error
+    if (this->sftpHandle == NULL)
+    {
+        const int rc = libssh2_session_last_errno(this->session);
+
+        if (rc == LIBSSH2_ERROR_SFTP_PROTOCOL)
+        {
+            uint64_t sftpErr = libssh2_sftp_last_error(this->sftpSession);
+
+            if (sftpErr == LIBSSH2_FX_NO_SUCH_FILE)
+                THROW_FMT(FileMissingError, STORAGE_ERROR_WRITE_MISSING, strZ(this->interface.name));
+            else
+            {
+                storageSftpEvalLibSsh2Error(
+                    rc, sftpErr, &FileOpenError, strNewFmt(STORAGE_ERROR_WRITE_OPEN, strZ(this->interface.name)), NULL);
+            }
+        }
+        else
+            THROW_FMT(FileOpenError, STORAGE_ERROR_WRITE_OPEN, strZ(this->interface.name));
+    }
+
+    // Update user/group owner
+    if (this->interface.user != NULL || this->interface.group != NULL)
```

Comment:
This code is unreachable outside of unit tests. Backups/archive-push will never pass a user or group name to storageNewWriteP().

Reply:
removed unreachable code, commit f25431f5b67377280f90996e8aead476d88c2b6a